packages/@aws-cdk/aws-s3/lib/bucket.ts

```diff
@@ -584,7 +593,7 @@ abstract class BucketBase extends Resource implements IBucket {
    * @param objectsKeyPattern Restrict the permission to a certain key pattern (default '*')
    */
   public grantWrite(identity: iam.IGrantable, objectsKeyPattern: any = '*') {
-    return this.grant(identity, perms.BUCKET_WRITE_ACTIONS, perms.KEY_WRITE_ACTIONS,
+    return this.grant(identity, this.writeActions, perms.KEY_WRITE_ACTIONS,
```

Comment:
I'm kinda expecting this change to break this:

https://github.com/aws/aws-cdk/blob/49a9116f4e58ddda44f9ff7faf4af5d16d24f744/packages/%40aws-cdk/aws-codepipeline-actions/lib/s3/deploy-action.ts#L112

if this field is set: 

https://github.com/aws/aws-cdk/blob/49a9116f4e58ddda44f9ff7faf4af5d16d24f744/packages/%40aws-cdk/aws-codepipeline-actions/lib/s3/deploy-action.ts#L78

I think it makes sense to add a method to explicitly opt in the old behavior, which we're going to need to use in at least the CodePipeline Action, and potentially other cases. 

It's also a lot more user friendly, for everyone that intentionally needs `PutObjectAcl`. It's a much better advisory to say:

"Replace `grantWrite` with `grantFullWrite`, because security"

Than to say:

"Replace `grantWrite` with `grant("s3:GetObject*", "s3:Get*", "s3:DeleteObject*", "s3:Abort*", "s3:PutObject*", ...)`, because security"

Reply:
I don't disagree, but do you maybe want to wait for our users to let us know if it breaks, instead of doing it in this PR? What if you're wrong, and it works? 🙂

Reply:
I'm proposing you test :P

Reply:
Did it for you:

![image](https://user-images.githubusercontent.com/524162/104165124-3f5db500-53f9-11eb-8630-027e0a8838a3.png)

Repro in case you need it:

```ts
import * as cpipe from '@aws-cdk/aws-codepipeline';
import { App, RemovalPolicy, SecretValue, Stack, StackProps } from '@aws-cdk/core';
import { Construct } from 'constructs';
import * as s3 from '@aws-cdk/aws-s3';
import * as actions from '../../lib';

class MyStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const bukkit = new s3.Bucket(this, 'Bukkit', {
      removalPolicy: RemovalPolicy.DESTROY,
    });

    const arto = new cpipe.Artifact();
    const pipe = new cpipe.Pipeline(this, 'Pipe');
    pipe.addStage({
      stageName: 'Source',
      actions: [
        new actions.GitHubSourceAction({
          actionName: 'GitHub',
          oauthToken: SecretValue.secretsManager('github-token'),
          output: arto,
          owner: 'rix0rrr',
          repo: 'kessel-run',
          branch: 'main',
        }),
      ],
    });
    pipe.addStage({
      stageName: 'PutErThere',
      actions: [
        new actions.S3DeployAction({
          actionName: 'InS3',
          bucket: bukkit,
          input: arto,
          accessControl: s3.BucketAccessControl.PUBLIC_READ,
        }),
      ],
    });
  }
}

const app = new App({
  context: {
    '@aws-cdk/aws-s3:grantWriteWithoutAcl': true,
  },
});
new MyStack(app, 'DeployStack', {
  env: {
    region: process.env.CDK_DEFAULT_REGION,
    account: process.env.CDK_DEFAULT_ACCOUNT,
  },
});
app.synth();
```

Reply:
Thanks Rico! I completely blanked on the fact that `accessControl` in `S3DeployAction` actually uses Bucket ACLs to achieve its functionality.

I confirmed that setting any `accessControl` does require `s3:PutObjectAcl`. Although it's possible to deploy publicly readable objects into a Bucket without using `accessControl` (you have to change the resource policy of the Bucket to allow everyone to read), we have to support `accessControl` as well.

I've added a `grantPutAcl()` to `IBucket`, and changed the `S3DeployAction` to call that method if the feature flag is set and `accessControl` was passed.

I only added `s3:PutObjectAcl` permissions in `grantPutAcl()` for now; contrary to the (I agree excellent) error message above, I was not able to get a situation where `s3:PutObjectVersionAcl` would be required (yes, I tried overwriting the same file multiple times, didn't cause see any issues with just `s3:PutObjectAcl`), so I left it out for now; we can always add it later.

Let me know if this works!

Reply:
> yes, I tried overwriting the same file multiple times

Did you try this with a versioned bucket?

Reply:
> > yes, I tried overwriting the same file multiple times
> 
> Did you try this with a versioned bucket?

Yep, I did. Exactly the same result (no error even though the Action's Role did not have `s3:PutObjectVersionAcl` permissions, the resulting object was readable publicly, multiple versions were created without problems).